examples/vision/autoencoder.py

```diff
@@ -0,0 +1,174 @@
+from tensorflow.keras.layers import Conv2D, MaxPooling2D, Input, UpSampling2D
+from tensorflow.keras.datasets import mnist
+
+
+def preprocess(dataset):
+    """
+    Normalizes the supplied dataset and reshapes it into the appropriate format.
+    """
+
+    dataset = dataset.astype("float32") / 255.0
+    dataset = np.reshape(dataset, (len(dataset), 28, 28, 1))
+    return dataset
+
+
+def noise(dataset):
+    """
+    Adds random noise to each image in the supplied dataset.
+    """
+
+    noise_factor = 0.4
+    noisy_dataset = dataset + noise_factor * np.random.normal(
+        loc=0.0, scale=1.0, size=dataset.shape
+    )
+
+    return np.clip(noisy_dataset, 0.0, 1.0)
+
+
+def display(dataset1, dataset2):
+    """
+    Displays ten random images from each one of the supplied datasets.
+    """
+
+    n = 10
+
+    indices = np.random.randint(len(dataset1), size=n)
+    images1 = dataset1[indices, :]
+    images2 = dataset2[indices, :]
+
+    plt.figure(figsize=(20, 4))
+    for i, (image1, image2) in enumerate(zip(images1, images2)):
+        ax = plt.subplot(2, n, i + 1)
+        plt.imshow(image1.reshape(28, 28))
+        plt.gray()
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+
+        ax = plt.subplot(2, n, i + 1 + n)
+        plt.imshow(image2.reshape(28, 28))
+        plt.gray()
+        ax.get_xaxis().set_visible(False)
+        ax.get_yaxis().set_visible(False)
+
+    plt.show()
+
+
+"""
+## Prepare de data
+"""
+
+# We only need images from the dataset. We will never use the target values so
+# we aren't going to load them.
+(train_dataset, _), (test_dataset, _) = mnist.load_data()
+
+# Normalize and reshape the data
+train_dataset = preprocess(train_dataset)
+test_dataset = preprocess(test_dataset)
+
+# Create a copy of the datasets with added noise
+noisy_train_dataset = noise(train_dataset)
+noisy_test_dataset = noise(test_dataset)
+
+# Display the train dataset and the version with added noise
+display(train_dataset, noisy_train_dataset)
+
+"""
+## Build the autoencoder
+
+We are going to use the functional API to build our convolutional autoencoder.
+"""
+
+input = Input(shape=(28, 28, 1))
+
+# Encoder
+x = Conv2D(32, (3, 3), activation="relu", padding="same")(input)
+x = MaxPooling2D((2, 2), padding="same")(x)
+x = Conv2D(32, (3, 3), activation="relu", padding="same")(x)
+x = MaxPooling2D((2, 2), padding="same")(x)
+
+# Decoder
+x = Conv2D(32, (3, 3), activation="relu", padding="same")(x)
+x = UpSampling2D((2, 2))(x)
+x = Conv2D(32, (3, 3), activation="relu", padding="same")(x)
+x = UpSampling2D((2, 2))(x)
+x = Conv2D(1, (3, 3), activation="sigmoid", padding="same")(x)
+
+# Autoencoder
+autoencoder = Model(input, x)
+autoencoder.compile(optimizer="adam", loss="binary_crossentropy")
+autoencoder.summary()
+
+"""
+Now we can train our autoencoder using the `train_dataset` as our `X` and the same
+dataset as our `y`. Noticed we are setting up the validation data using the same format.
```

Comment:
Suggestion:
- Refactor
- Fix type ("Notice")

```suggestion
Now we can train the autoencoder using `train_dataset` as our input `X` and the same
dataset as our `y`. Notice that we are setting up the validation data using the same format.
```